src/hdmf_zarr/zarr_utils.py

```diff
@@ -0,0 +1,289 @@
+import warnings
+import os
+import logging
+
+#from Zarrpy import Group, Dataset, RegionReference, Reference, special_dtype
+#from Zarrpy import filters as Zarrpy_filter
+from zarr import Array as ZarrArray
+
+from hdmf.array import Array
+from hdmf.query import HDMFDataset, ReferenceResolver, ContainerResolver, BuilderResolver
+from hdmf.utils import docval, getargs, popargs, get_docval
+
+
+class ZarrDataset(HDMFDataset):
+    @docval({'name': 'dataset', 'type': (np.ndarray, ZarrArray, Array), 'doc': 'the Zarr file lazily evaluate'},
+            {'name': 'io', 'type': 'ZarrIO', 'doc': 'the IO object that was used to read the underlying dataset'})
+    def __init__(self, **kwargs):
+        self.__io = popargs('io', kwargs)
+        super().__init__(**kwargs)
+
+    @property
+    def io(self):
+        return self.__io
+
+    @property
+    def shape(self):
+        return self.dataset.shape
+
+
+class DatasetOfReferences(ZarrDataset, ReferenceResolver, metaclass=ABCMeta):
+    """
+    An extension of the base ReferenceResolver class to add more abstract methods for
+    subclasses that will read Zarr references
+    """
+
+    @abstractmethod
+    def get_object(self, zarr_obj):
+        """
+        A class that maps an Zarr object to a Builder or Container
+        """
+        pass
+
+    def invert(self):
+        """
+        Return an object that defers reference resolution
+        but in the opposite direction.
+        """
+        if not hasattr(self, '__inverted'):
+            cls = self.get_inverse_class()
+            docval = get_docval(cls.__init__)
+            kwargs = dict()
+            for arg in docval:
+                kwargs[arg['name']] = getattr(self, arg['name'])
+            self.__inverted = cls(**kwargs)
+        return self.__inverted
+
+    def _get_ref(self, ref):
+        # return self.get_object(self.dataset.file[ref])
+        # breakpoint()
+        name, zarr_obj = self.io.resolve_ref(ref) # ref is a json dict containing the path to the object
+        return self.get_object(zarr_obj)
+
+    def __iter__(self):
+        for ref in super().__iter__():
+            yield self._get_ref(ref)
+
+    def __next__(self):
+        return self._get_ref(super().__next__())
+
+
+class BuilderResolverMixin(BuilderResolver):# refactor to backend/utils.py
+    """
+    A mixin for adding to Zarr reference-resolving types
+    the get_object method that returns Builders
+    """
+
+    def get_object(self, zarr_obj):
+        """
+        A class that maps an Zarr object to a Builder
+        """
+        return self.io.get_builder(zarr_obj)
+
+
+class ContainerResolverMixin(ContainerResolver): # refactor to backend/utils.py
+    """
+    A mixin for adding to Zarr reference-resolving types
+    the get_object method that returns Containers
+    """
+
+    def get_object(self, zarr_obj):
+        """
+        A class that maps an Zarr object to a Container
+        """
+        return self.io.get_container(zarr_obj)
```

Comment:
I think `BuilderResolverMixin` and `ContainerResolverMixin` are not implementing any Zarr-specific functionality. I think we may be able to just import these from HDMF. As the comment says, for import we may want to move them to `backend/utils.py` in HDMF instead